From 30b38bb01252a97a3271318e7e5945169ee58b4a Mon Sep 17 00:00:00 2001
From: Wenyou Yang <wenyou.yang@atmel.com>
Date: Mon, 9 Feb 2015 12:00:08 +0800
Subject: [PATCH] driver: pm: move the pm stuff code from the board file to the
 new file

Create a new file to contail the pm stuff, move the stuff from the board file to it,
to decrease the the duplicated code.

In the meanwhile,
	- Rename CONFIG_PM -->CONFIG_SUPPORT_PM
	- Rename CONFIG_PM_EXTERNAL_DEVICES --> CONFIG_PM

Signed-off-by: Wenyou Yang <wenyou.yang@atmel.com>
---
 Config.in                                 | 16 +++++-----
 board/Config.in                           |  2 +-
 board/sama5d3_xplained/Config.in.board    |  2 +-
 board/sama5d3_xplained/sama5d3_xplained.c | 22 ++++---------
 board/sama5d4_xplained/Config.in.board    |  2 +-
 board/sama5d4_xplained/sama5d4_xplained.c | 27 ++++------------
 board/sama5d4ek/Config.in.board           |  2 +-
 board/sama5d4ek/sama5d4ek.c               | 30 ++++--------------
 driver/driver.mk                          |  1 +
 driver/driver_cpp.mk                      |  4 +--
 driver/pm.c                               | 52 +++++++++++++++++++++++++++++++
 include/board.h                           |  3 ++
 include/pm.h                              | 33 ++++++++++++++++++++
 main.c                                    |  6 ++++
 14 files changed, 127 insertions(+), 75 deletions(-)
 create mode 100644 driver/pm.c
 create mode 100644 include/pm.h

diff --git a/Config.in b/Config.in
index 3809a8e..fd80dab 100644
--- a/Config.in
+++ b/Config.in
@@ -366,18 +366,18 @@ config CONFIG_REDIRECT_ALL_INTS_AIC
 
 endmenu
 
-config CONFIG_PM_EXTERNAL_DEVICES
+config CONFIG_PM
 	bool "Power Management Options"
-	depends on CONFIG_PM
-	default n
+	depends on CONFIG_SUPPORT_PM
+	default y
 	help
-	  This interface let you select make the external devices on board
-	  enter the power down mode at startup, such as Codec wm8904, HDMI SiI9022ACUN
-	  and two PHYs KSZ8081RNB. Its purpose is to make sure the external devices
+	  This interface let you select the external devices on the board to
+	  enter the power down mode at startup, such as Audio Codec(wm8904), HDMI(SiI9022ACUN)
+	  and ethernet PHY(KSZ8081RNB). Its purpose is to make sure the external devices
 	  which will not be touched in the kernel are set the proper power state.
 
-menu "Select the Devices to enter the Power-Down mode"
-	depends on CONFIG_PM_EXTERNAL_DEVICES
+menu "Select the Devices to the Low-power mode"
+	depends on CONFIG_PM
 
 config CONFIG_MAC0_PHY
 	bool "MAC0 PHYs"
diff --git a/board/Config.in b/board/Config.in
index 49327c2..7681bfc 100644
--- a/board/Config.in
+++ b/board/Config.in
@@ -369,7 +369,7 @@ config CONFIG_HAS_PMIC_ACT8865
 	bool
 	default n
 
-config CONFIG_PM
+config CONFIG_SUPPORT_PM
 	bool
 	default n
 
diff --git a/board/sama5d3_xplained/Config.in.board b/board/sama5d3_xplained/Config.in.board
index 13cdfff..c36ded2 100644
--- a/board/sama5d3_xplained/Config.in.board
+++ b/board/sama5d3_xplained/Config.in.board
@@ -14,7 +14,7 @@ config CONFIG_SAMA5D3_XPLAINED
 	select ALLOW_PIO3
 	select CPU_HAS_PMECC
 	select CONFIG_MMC_SUPPORT
-	select CONFIG_PM
+	select CONFIG_SUPPORT_PM
 	select CONFIG_HAS_EHT0_PHY
 	select CONFIG_HAS_EHT1_PHY
 	select CONFIG_HAS_PMIC_ACT8865
diff --git a/board/sama5d3_xplained/sama5d3_xplained.c b/board/sama5d3_xplained/sama5d3_xplained.c
index 690b38f..e2e6360 100644
--- a/board/sama5d3_xplained/sama5d3_xplained.c
+++ b/board/sama5d3_xplained/sama5d3_xplained.c
@@ -234,8 +234,6 @@ static void at91_special_pio_output_low(void)
 }
 #endif
 
-#if defined(CONFIG_PM_EXTERNAL_DEVICES)
-#if defined(CONFIG_MACB)
 #if defined(CONFIG_MAC0_PHY)
 static void gmac_hw_init(void)
 {
@@ -264,7 +262,8 @@ static void emac_hw_init(void)
 }
 #endif
 
-static int phys_enter_power_down(void)
+#ifdef CONFIG_MACB
+int phys_enter_power_down(void)
 {
 	struct mii_bus macb_mii_bus;
 
@@ -304,8 +303,7 @@ static int phys_enter_power_down(void)
 
 	return 0;
 }
-#endif	/* #if defined(CONFIG_MACB) */
-#endif	/* #if defined(CONFIG_PM_EXTERNAL_DEVICES) */
+#endif
 
 #if defined(CONFIG_TWI0)
 unsigned int at91_twi0_hw_init(void)
@@ -377,7 +375,8 @@ static int sama5d3ek_act8865_set_reg_voltage(void)
 }
 #endif
 
-static void at91_disable_smd_clock(void)
+#if defined(CONFIG_PM)
+void at91_disable_smd_clock(void)
 {
 	/*
 	 * set pin DIBP to pull-up and DIBN to pull-down
@@ -390,6 +389,7 @@ static void at91_disable_smd_clock(void)
 	pmc_disable_periph_clock(AT91C_ID_SMD);
 	pmc_disable_system_clock(AT91C_PMC_SMDCK);
 }
+#endif
 
 #ifdef CONFIG_HW_INIT
 void hw_init(void)
@@ -419,9 +419,6 @@ void hw_init(void)
 	at91_special_pio_output_low();
 #endif
 
-	/* Disable the software modem clock */
-	at91_disable_smd_clock();
-
 	/* Init timer */
 	timer_init();
 
@@ -453,13 +450,6 @@ void hw_init(void)
 		while (1)
 			;
 #endif
-
-#ifdef CONFIG_PM_EXTERNAL_DEVICES
-#ifdef CONFIG_MACB
-	/* Make PHYs to power down mode */
-	phys_enter_power_down();
-#endif
-#endif	/* #ifdef CONFIG_PM_EXTERNAL_DEVICES */
 }
 #endif /* #ifdef CONFIG_HW_INIT */
 
diff --git a/board/sama5d4_xplained/Config.in.board b/board/sama5d4_xplained/Config.in.board
index d54356e..8ce356e 100644
--- a/board/sama5d4_xplained/Config.in.board
+++ b/board/sama5d4_xplained/Config.in.board
@@ -19,7 +19,7 @@ config CONFIG_SAMA5D4_XPLAINED
 	select CONFIG_MMC_SUPPORT
 	select CONFIG_TRUSTZONE
 	select CONFIG_HAS_PMIC_ACT8865
-	select CONFIG_PM
+	select CONFIG_SUPPORT_PM
 	select CONFIG_HAS_EHT0_PHY
 	select CONFIG_HAS_HDMI
 	select SUPPORT_BUS_SPEED_176MHZ
diff --git a/board/sama5d4_xplained/sama5d4_xplained.c b/board/sama5d4_xplained/sama5d4_xplained.c
index 74e0bf8..8b2964b 100644
--- a/board/sama5d4_xplained/sama5d4_xplained.c
+++ b/board/sama5d4_xplained/sama5d4_xplained.c
@@ -49,8 +49,6 @@
 #include "matrix.h"
 #include "twi.h"
 #include "act8865.h"
-#include "hdmi_SiI9022.h"
-#include "wm8904.h"
 #include "macb.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
@@ -549,7 +547,8 @@ static int sama5d4ek_act8865_set_reg_voltage(void)
 }
 #endif
 
-static void at91_disable_smd_clock(void)
+#if defined(CONFIG_PM)
+void at91_disable_smd_clock(void)
 {
 	/*
 	 * set pin DIBP to pull-up and DIBN to pull-down
@@ -562,6 +561,7 @@ static void at91_disable_smd_clock(void)
 	pmc_disable_periph_clock(AT91C_ID_SMD);
 	pmc_disable_system_clock(AT91C_PMC_SMDCK);
 }
+#endif
 
 static void SiI9022_hw_reset(void)
 {
@@ -571,8 +571,6 @@ static void SiI9022_hw_reset(void)
 	pio_set_gpio_output(CONFIG_SYS_HDMI_RESET_PIN, 1);
 }
 
-#if defined(CONFIG_PM_EXTERNAL_DEVICES)
-#if defined(CONFIG_MACB)
 #if defined(CONFIG_MAC0_PHY)
 static void gmac0_hw_init(void)
 {
@@ -587,7 +585,8 @@ static void gmac0_hw_init(void)
 }
 #endif
 
-static int phys_enter_power_down(void)
+#ifdef CONFIG_MACB
+int phys_enter_power_down(void)
 {
 	struct mii_bus macb_mii_bus;
 
@@ -610,8 +609,7 @@ static int phys_enter_power_down(void)
 
 	return 0;
 }
-#endif	/* #if defined(CONFIG_MACB) */
-#endif	/* #if defined(CONFIG_PM_EXTERNAL_DEVICES) */
+#endif
 
 #ifdef CONFIG_HW_INIT
 void hw_init(void)
@@ -673,9 +671,6 @@ void hw_init(void)
 	/* Reset HDMI SiI9022 */
 	SiI9022_hw_reset();
 
-	/* Disable software modem device's clock */
-	at91_disable_smd_clock();
-
 #ifdef CONFIG_TWI
 	twi_init();
 #endif
@@ -693,16 +688,6 @@ void hw_init(void)
 			;
 #endif
 
-#ifdef CONFIG_PM_EXTERNAL_DEVICES
-#ifdef CONFIG_HDMI
-	SiI9022_enter_power_state_D3_Cold();
-#endif
-#ifdef CONFIG_MACB
-	/* Make PHYs to power down mode */
-	phys_enter_power_down();
-#endif
-#endif	/* #ifdef CONFIG_PM_EXTERNAL_DEVICES */
-
 #ifdef CONFIG_USER_HW_INIT
 	hw_init_hook();
 #endif
diff --git a/board/sama5d4ek/Config.in.board b/board/sama5d4ek/Config.in.board
index b8d9aa9..b8ab5ae 100644
--- a/board/sama5d4ek/Config.in.board
+++ b/board/sama5d4ek/Config.in.board
@@ -19,7 +19,7 @@ config CONFIG_SAMA5D4EK
 	select CONFIG_MMC_SUPPORT
 	select CONFIG_TRUSTZONE
 	select CONFIG_HAS_PMIC_ACT8865
-	select CONFIG_PM
+	select CONFIG_SUPPORT_PM
 	select CONFIG_HAS_EHT0_PHY
 	select CONFIG_HAS_EHT1_PHY
 	select CONFIG_HAS_AUDIO_CODEC
diff --git a/board/sama5d4ek/sama5d4ek.c b/board/sama5d4ek/sama5d4ek.c
index 0d43114..7e8230b 100644
--- a/board/sama5d4ek/sama5d4ek.c
+++ b/board/sama5d4ek/sama5d4ek.c
@@ -50,8 +50,6 @@
 #include "matrix.h"
 #include "twi.h"
 #include "act8865.h"
-#include "hdmi_SiI9022.h"
-#include "wm8904.h"
 #include "macb.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
@@ -561,7 +559,8 @@ static int sama5d4ek_act8865_set_reg_voltage(void)
 }
 #endif
 
-static void at91_disable_smd_clock(void)
+#if defined(CONFIG_PM)
+void at91_disable_smd_clock(void)
 {
 	/*
 	 * set pin DIBP to pull-up and DIBN to pull-down
@@ -574,6 +573,7 @@ static void at91_disable_smd_clock(void)
 	pmc_disable_periph_clock(AT91C_ID_SMD);
 	pmc_disable_system_clock(AT91C_PMC_SMDCK);
 }
+#endif
 
 static void SiI9022_hw_reset(void)
 {
@@ -583,8 +583,6 @@ static void SiI9022_hw_reset(void)
 	pio_set_gpio_output(CONFIG_SYS_HDMI_RESET_PIN, 1);
 }
 
-#if defined(CONFIG_PM_EXTERNAL_DEVICES)
-#if defined(CONFIG_MACB)
 #if defined(CONFIG_MAC0_PHY)
 static void gmac0_hw_init(void)
 {
@@ -613,7 +611,8 @@ static void gmac1_hw_init(void)
 }
 #endif
 
-static int phys_enter_power_down(void)
+#ifdef CONFIG_MACB
+int phys_enter_power_down(void)
 {
 	struct mii_bus macb_mii_bus;
 
@@ -653,8 +652,7 @@ static int phys_enter_power_down(void)
 
 	return 0;
 }
-#endif	/* #if defined(CONFIG_MACB) */
-#endif	/* #if defined(CONFIG_PM_EXTERNAL_DEVICES) */
+#endif
 
 #ifdef CONFIG_HW_INIT
 void hw_init(void)
@@ -719,9 +717,6 @@ void hw_init(void)
 	/* Reset HDMI SiI9022 */
 	SiI9022_hw_reset();
 
-	/* Disable software modem device's clock */
-	at91_disable_smd_clock();
-
 #ifdef CONFIG_TWI
 	twi_init();
 #endif
@@ -739,19 +734,6 @@ void hw_init(void)
 			;
 #endif
 
-#ifdef CONFIG_PM_EXTERNAL_DEVICES
-#ifdef CONFIG_HDMI
-	SiI9022_enter_power_state_D3_Cold();
-#endif
-#ifdef CONFIG_WM8904
-	wm8904_enter_low_power();
-#endif
-#ifdef CONFIG_MACB
-	/* Make PHYs to power down mode */
-	phys_enter_power_down();
-#endif
-#endif	/* #ifdef CONFIG_PM_EXTERNAL_DEVICES */
-
 #ifdef CONFIG_USER_HW_INIT
 	hw_init_hook();
 #endif
diff --git a/driver/driver.mk b/driver/driver.mk
index 02308ef..232094f 100644
--- a/driver/driver.mk
+++ b/driver/driver.mk
@@ -51,6 +51,7 @@ COBJS-$(CONFIG_ENTER_NWD)	+= $(DRIVERS_SRC)/at91_rstc.o
 COBJS-$(CONFIG_ENTER_NWD)	+= $(DRIVERS_SRC)/monitor/mon_switch.o
 COBJS-$(CONFIG_ENTER_NWD)	+= $(DRIVERS_SRC)/monitor/mon_vectors.o
 
+COBJS-$(CONFIG_PM)	+= $(DRIVERS_SRC)/pm.o
 COBJS-$(CONFIG_TWI)	+= $(DRIVERS_SRC)/at91_twi.o
 COBJS-$(CONFIG_ACT8865)	+= $(DRIVERS_SRC)/act8865.o
 COBJS-$(CONFIG_MACB)	+= $(DRIVERS_SRC)/macb.o
diff --git a/driver/driver_cpp.mk b/driver/driver_cpp.mk
index 29e14f0..d04e9b8 100644
--- a/driver/driver_cpp.mk
+++ b/driver/driver_cpp.mk
@@ -221,8 +221,8 @@ ifeq ($(CONFIG_ACT8865), y)
 CPPFLAGS += -DCONFIG_ACT8865
 endif
 
-ifeq ($(CONFIG_PM_EXTERNAL_DEVICES), y)
-CPPFLAGS += -DCONFIG_PM_EXTERNAL_DEVICES
+ifeq ($(CONFIG_PM), y)
+CPPFLAGS += -DCONFIG_PM
 endif
 
 ifeq ($(CONFIG_MACB), y)
diff --git a/driver/pm.c b/driver/pm.c
new file mode 100644
index 0000000..56484d5
--- /dev/null
+++ b/driver/pm.c
@@ -0,0 +1,52 @@
+/* ----------------------------------------------------------------------------
+ *         ATMEL Microcontroller Software Support
+ * ----------------------------------------------------------------------------
+ * Copyright (c) 2015, Atmel Corporation
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the disclaiimer below.
+ *
+ * Atmel's name may not be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * DISCLAIMER: THIS SOFTWARE IS PROVIDED BY ATMEL "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT ARE
+ * DISCLAIMED. IN NO EVENT SHALL ATMEL BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "board.h"
+#include "twi.h"
+#include "wm8904.h"
+#include "hdmi_SiI9022.h"
+#include "macb.h"
+
+void at91_board_pm(void)
+{
+	at91_disable_smd_clock();
+
+#ifdef CONFIG_TWI
+	if (!twi_init_done)
+		twi_init();
+#endif
+
+#ifdef CONFIG_HDMI
+	SiI9022_enter_power_state_D3_Cold();
+#endif
+#ifdef CONFIG_WM8904
+	wm8904_enter_low_power();
+#endif
+#ifdef CONFIG_MACB
+	phys_enter_power_down();
+#endif
+}
diff --git a/include/board.h b/include/board.h
index 0c45e05..9ed583c 100644
--- a/include/board.h
+++ b/include/board.h
@@ -103,4 +103,7 @@ extern unsigned int at91_twi1_hw_init(void);
 extern unsigned int at91_twi2_hw_init(void);
 extern unsigned int at91_twi3_hw_init(void);
 
+extern int phys_enter_power_down(void);
+extern void at91_disable_smd_clock(void);
+
 #endif /* #ifndef __BOARD_H__ */
diff --git a/include/pm.h b/include/pm.h
new file mode 100644
index 0000000..ee57b47
--- /dev/null
+++ b/include/pm.h
@@ -0,0 +1,33 @@
+/* ----------------------------------------------------------------------------
+ *         ATMEL Microcontroller Software Support
+ * ----------------------------------------------------------------------------
+ * Copyright (c) 2015, Atmel Corporation
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the disclaiimer below.
+ *
+ * Atmel's name may not be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * DISCLAIMER: THIS SOFTWARE IS PROVIDED BY ATMEL "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT ARE
+ * DISCLAIMED. IN NO EVENT SHALL ATMEL BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __PM_H__
+#define __PM_H__
+
+extern void at91_board_pm(void);
+
+#endif
diff --git a/main.c b/main.c
index f91dee6..fbc9837 100644
--- a/main.c
+++ b/main.c
@@ -38,6 +38,7 @@
 #include "string.h"
 #include "board_hw_info.h"
 #include "tz_utils.h"
+#include "pm.h"
 
 extern int load_kernel(struct image_info *img_info);
 
@@ -151,6 +152,11 @@ int main(void)
 	/* Load board hw informaion */
 	load_board_hw_info();
 #endif
+
+#ifdef CONFIG_PM
+	at91_board_pm();
+#endif
+
 	init_loadfunction();
 
 	ret = (*load_image)(&image);
-- 
1.9.1

